### PR TITLE
Export as DLL

### PR DESCRIPTION
**Pull request by JunielKatarn**
- Adds DLL-generating project.
- Updates the NuGet spec so import lib and binaries are correctly laid out.

### CODE REVIEW

File: bin/Debugger/ChakraCore.Debugger.def [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+; Protocol Handler
+JsDebugProtocolHandlerCreate
+JsDebugProtocolHandlerDestroy
+JsDebugProtocolHandlerConnect
+JsDebugProtocolHandlerDisconnect
+JsDebugProtocolHandlerSendCommand
+JsDebugProtocolHandlerWaitForDebugger
+JsDebugProtocolHandlerProcessCommandQueue
+JsDebugProtocolHandlerSetCommandQueueCallback
+JsDebugProtocolHandlerCreateConsoleObject
```
Comment (NikoAri) [inline bin/Debugger/ChakraCore.Debugger.def]: >dlerCre [](start = 18, length = 7)

(nit) please alphabetize

File: NuGet/Microsoft.ChakraCore.Debugger.nuspec [hunk 2]
```diff
@@ -17,15 +17,15 @@
   </metadata>
 
   <files>
-    <file src="..\lib\Debugger.Service\ChakraDebugService.h" target="include" />
-    <file src="..\lib\Debugger.ProtocolHandler\ChakraDebugProtocolHandler.h" target="include" />
+    <file src="..\lib\Debugger.Service\ChakraDebugService.h" target="build\native\include" />
```
Comment (NikoAri) [inline NuGet/Microsoft.ChakraCore.Debugger.nuspec]: >build [](start = 69, length = 5)

Why is this folder called "build"? It probably should be called "inc", "publicinc", or something like that.

Also, why do we need to specify "native"? Is it to differentiate from some javascript/typescript definition files?

File: NuGet/Microsoft.ChakraCore.Debugger.nuspec [hunk 3]
```diff
@@ -17,15 +17,15 @@
...
+    <file src="..\lib\Debugger.ProtocolHandler\ChakraDebugProtocolHandler.h" target="build\native\include" />
 
-    <file src="..\build\bin\x64\Debug\*" target="lib\x64\Debug" exclude="**\*.ilk;**\*ChakraCore.dll;**\*ChakraCore.pdb" />
-    <file src="..\build\bin\x64\Release\*" target="lib\x64\Release" exclude="**\*.ilk;**\*ChakraCore.dll;**\*ChakraCore.pdb" />
-    <file src="..\build\bin\x86\Debug\*" target="lib\x86\Debug" exclude="**\*.ilk;**\*ChakraCore.dll;**\*ChakraCore.pdb" />
-    <file src="..\build\bin\x86\Release\*" target="lib\x86\Release" exclude="**\*.ilk;**\*ChakraCore.dll;**\*ChakraCore.pdb" />
-    <file src="..\build\bin\ARM\Debug\*" target="lib\ARM\Debug" exclude="**\*.ilk;**\*ChakraCore.dll;**\*ChakraCore.pdb" />
-    <file src="..\build\bin\ARM\Release\*" target="lib\ARM\Release" exclude="**\*.ilk;**\*ChakraCore.dll;**\*ChakraCore.pdb" />
+    <file src="..\build\bin\x64\Debug\ChakraCore.Debugger.???" target="lib\native\x64\Debug" exclude="**\*.ilk;**\*.exp" />
```
Comment (NikoAri) [inline NuGet/Microsoft.ChakraCore.Debugger.nuspec]: >native [](start = 75, length = 6)

Similar to previous comment, why specify "native"? What is the recommended folder structure?

File: NuGet/Microsoft.ChakraCore.Debugger.targets [hunk 4]
```diff
@@ -1,25 +1,31 @@
...
       </AdditionalDependencies>
     </Link>
   </ItemDefinitionGroup>
 
+  <ItemGroup Condition="'$(Platform)' != 'Win32'">
+    <ReferenceCopyLocalPaths Include="$(MSBuildThisFileDirectory)..\..\lib\native\$(Platform)\$(Configuration)\ChakraCore.Debugger.dll" />
+    <ReferenceCopyLocalPaths Include="$(MSBuildThisFileDirectory)..\..\lib\native\$(Platform)\$(Configuration)\ChakraCore.Debugger.pdb" />
+  </ItemGroup>
+  <ItemGroup Condition="'$(Platform)' == 'Win32'">
+    <ReferenceCopyLocalPaths Include="$(MSBuildThisFileDirectory)..\..\lib\native\x86\$(Configuration)\ChakraCore.Debugger.dll" />
```
Comment (NikoAri) [inline NuGet/Microsoft.ChakraCore.Debugger.targets]: >x86 [](start = 82, length = 3)

Why not use $(PlatformTarget) macro instead? That way you can get rid of the whole Win32 "Condition" node.

File: bin/Debugger/ChakraCore.Debugger.vcxproj [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  <ItemGroup>
+    <ClCompile Include="stdafx.cpp" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="ChakraCore.Debugger.def" />
+    <None Include="packages.config" />
+  </ItemGroup>
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
+  <ImportGroup Label="ExtensionTargets">
+    <Import Project="..\..\packages\Microsoft.ChakraCore.vc140.1.11.10\build\native\Microsoft.ChakraCore.vc140.targets" Condition="Exists('..\..\packages\Microsoft.ChakraCore.vc140.1.11.10\build\native\Microsoft.ChakraCore.vc140.targets')" />
```
Comment (NikoAri) [inline bin/Debugger/ChakraCore.Debugger.vcxproj]: >s\M [](start = 34, length = 3)

(nit) please alphabetize if possible

Comment (NikoAri) [inline bin/Debugger/ChakraCore.Debugger.vcxproj]: >Condition [](start = 120, length = 9)

Is this "Condition" needed? What happens if this condition fails?

**Comment by NikoAri** [comment]
> <?xml version="1.0" encoding="utf-8"?>

This project is located in "bin" folder. Is that on purpose? "bin" usually implies target or build folders

---
Refers to: bin/Debugger/ChakraCore.Debugger.vcxproj:1 in bf6b182. [](commit_id = bf6b182f88eb28b74056616f0f3511be672cdde2, deletion_comment = False)

File: bin/Debugger/ChakraCore.Debugger.vcxproj [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
+  </PropertyGroup>
+  <PropertyGroup Label="Configuration">
+    <ConfigurationType>DynamicLibrary</ConfigurationType>
+  </PropertyGroup>
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
+  <Import Project="$(ProjectDir)..\..\PropertySheets\Chakra.Cpp.props" />
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
+  <ImportGroup Label="ExtensionSettings">
+  </ImportGroup>
```
Comment (NikoAri) [inline bin/Debugger/ChakraCore.Debugger.vcxproj]: (nit) remove this line for consistency.

Line 43 too.

Comment (NikoAri) [reply bin/Debugger/ChakraCore.Debugger.vcxproj]: And line 48 

---
In reply to: [324775137](https://github.com/microsoft/ChakraCore-Debugger/pull/81#discussion_r324775137) [](ancestors = 324775137)

File: bin/Debugger/ChakraCore.Debugger.vcxproj [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  <ImportGroup Label="Shared">
+  </ImportGroup>
+  <ImportGroup Label="PropertySheets">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+  </ImportGroup>
+  <PropertyGroup Label="UserMacros" />
+  <PropertyGroup />
+  <ItemDefinitionGroup>
+    <ClCompile>
+      <!-- <PrecompiledHeader>Use</PrecompiledHeader> -->
```
Comment (NikoAri) [inline bin/Debugger/ChakraCore.Debugger.vcxproj]: >PrecompiledHeader [](start = 12, length = 17)

Is this meant to be commented out? If so, remove the line and remove stdafx.cpp file

File: bin/Debugger/ChakraCore.Debugger.vcxproj [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      <PreprocessorDefinitions>%(PreprocessorDefinitions)</PreprocessorDefinitions>
+    </ClCompile>
+    <Link>
+      <SubSystem>Windows</SubSystem>
+    </Link>
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup>
+    <ClCompile>
```
Comment (NikoAri) [inline bin/Debugger/ChakraCore.Debugger.vcxproj]: ><ClCompile> [](start = 4, length = 11)

(nit) why are there 2 indentical definitions for Compiling? Here and line 50?

File: NuGet/Microsoft.ChakraCore.Debugger.nuspec (hunk 2 above)
Comment (JunielKatarn) [reply NuGet/Microsoft.ChakraCore.Debugger.nuspec]: I'm adhering to the convention used by the ChakraCore package.

File: NuGet/Microsoft.ChakraCore.Debugger.nuspec (hunk 3 above)
Comment (JunielKatarn) [reply NuGet/Microsoft.ChakraCore.Debugger.nuspec]: There is no absolute standard for NuGet native packages, though there are different recommendations.

I'm adhering to the convention followed by ChakraCore itself.

File: NuGet/Microsoft.ChakraCore.Debugger.targets (hunk 4 above)
Comment (JunielKatarn) [reply NuGet/Microsoft.ChakraCore.Debugger.targets]: I tried using that variable a while ago in RNW and failed.
Giving it another try...

File: bin/Debugger/ChakraCore.Debugger.vcxproj (hunk 5 above)
Comment (JunielKatarn) [reply bin/Debugger/ChakraCore.Debugger.vcxproj]: It's added by Visual Studio NuGet tooling.

Comment (JunielKatarn) [reply bin/Debugger/ChakraCore.Debugger.vcxproj]: This frequently gets disrupted by the VS NuGet tooling.
Won't fix, for it doesn't have a practical effect on the build process.

File: bin/Debugger/ChakraCore.Debugger.vcxproj (hunk 8 above)
Comment (JunielKatarn) [reply bin/Debugger/ChakraCore.Debugger.vcxproj]: My mistake. They are clearly redundant.

File: NuGet/Microsoft.ChakraCore.Debugger.targets (hunk 4 above)
Comment (JunielKatarn) [reply NuGet/Microsoft.ChakraCore.Debugger.targets]: `PlatformTarget` works in this instance, so applying the change.

Note, there will be instances in the MSBuild files where this fix won't work due to some legacy path conventions in PROPS and TARGETS files.

**Comment by JunielKatarn** [comment]
> 🕐

All addressed.

**Comment by akrantz** [comment]
@JunielKatarn You need to resolve merge conflicts with the master branch due to the other pull requests which were merged...

**Comment by JunielKatarn** [comment]
Thanks!

**Pull request merged**